drivers/serial/uart_stm32.c

```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+	LL_USART_DisableIT_RXNE(UartInstance);
+
+	data->dma_rx.blk_cfg.block_size = buf_size;
+	data->dma_rx.blk_cfg.dest_address = (uint32_t)data->dma_rx.buffer;
+
+	ret = dma_config(data->dev_dma_rx, data->dma_rx.dma_channel,
+				&data->dma_rx.dma_cfg);
+
+	if (ret != 0) {
+		LOG_ERR("UART ERR: RX DMA config failed!");
+		return -EINVAL;
+	}
+
+	if (dma_start(data->dev_dma_rx, data->dma_rx.dma_channel)) {
+		LOG_ERR("UART ERR: RX DMA start failed!");
+		return -EFAULT;
+	}
+
+	/* Enable RX DMA requests */
+	uart_stm32_dma_rx_enable(dev);
+
+	/* Enable IRQ IDLE to define the end of a
+	 * RX DMA transaction.
+	 */
+	LL_USART_ClearFlag_IDLE(UartInstance);
+	LL_USART_EnableIT_IDLE(UartInstance);
+
+	LL_USART_EnableIT_ERROR(UartInstance);
+
+	/* Request next buffer */
+	async_evt_rx_buf_request(data);
+
+	LOG_DBG("async rx enabled");
+
+	return ret;
+}
+
+static int uart_stm32_async_tx_abort(const struct device *dev)
+{
+	struct uart_stm32_data *data = DEV_DATA(dev);
+	size_t tx_buffer_length = data->dma_tx.buffer_length;
+	struct dma_status stat;
+
+	if (tx_buffer_length == 0) {
+		return -EFAULT;
+	}
+
+	k_delayed_work_cancel(&data->dma_tx.timeout_work);
+	if (!dma_get_status(data->dev_dma_tx,
+				data->dma_tx.dma_channel, &stat)) {
+		data->dma_tx.counter = tx_buffer_length - stat.pending_length;
+	}
+
+	dma_stop(data->dev_dma_tx, data->dma_tx.dma_channel);
+	async_evt_tx_abort(data);
+
+	return 0;
+}
+
+static void uart_stm32_async_rx_timeout(struct k_work *work)
+{
+	LOG_DBG("rx timeout");
+
+	struct uart_dma_stream *rx_stream = CONTAINER_OF(work,
+			struct uart_dma_stream, timeout_work);
+
+	struct uart_stm32_data *data = CONTAINER_OF(rx_stream,
+			struct uart_stm32_data, dma_rx);
+
+	const struct device *dev = data->uart_dev;
+
+	if (data->dma_rx.counter == data->dma_rx.buffer_length) {
+		uart_stm32_async_rx_disable(dev);
+	}
+
+	uart_stm32_dma_rx_flush(dev);
+}
+
+static void uart_stm32_async_tx_timeout(struct k_work *work)
+{
+	struct uart_dma_stream *tx_stream = CONTAINER_OF(work,
+			struct uart_dma_stream, timeout_work);
+
+	struct uart_stm32_data *data = CONTAINER_OF(tx_stream,
+			struct uart_stm32_data, dma_tx);
+
+	const struct device *dev = data->uart_dev;
+
+	uart_stm32_async_tx_abort(dev);
+
+	LOG_DBG("tx: async timeout");
+}
+
+static int uart_stm32_async_rx_buf_rsp(const struct device *dev, uint8_t *buf,
+				       size_t len)
+{
+	struct uart_stm32_data *data = DEV_DATA(dev);
+
+	LOG_DBG("replace buffer (%d)", len);
+	data->rx_next_buffer = buf;
+	data->rx_next_buffer_len = len;
+
+	return 0;
+}
+
+static int uart_stm32_async_init(const struct device *dev)
+{
+	struct uart_stm32_data *data = DEV_DATA(dev);
+	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+
+	data->uart_dev = dev;
+
+	if (data->dma_rx.dma_name != NULL) {
+		data->dev_dma_rx = device_get_binding(data->dma_rx.dma_name);
+		if (data->dev_dma_rx == NULL) {
+			LOG_ERR("%s device not found", data->dma_rx.dma_name);
+			return -ENODEV;
+		}
+	}
+
```

Comment:
No error if `data->dma_rx.dma_name == NULL` ?

Reply:
No, since it is called in the init function of the async_api which is called in the init of the uart driver. It is valid to not include any dma in uarts that don't want that.